### DISCUSSION TITLE
Why does integrals from r_outcore and kernel in ao2omo vary?

### DISCUSSION THREAD

**Discussion by ironhide210** [pyscf/pyscf#2644] [Q&A]
So when I generate the OOVV integral for GHF  using r_outcore in ao2omo, and if compare with the OOVV integral generated using the inbuilt pyscf.GCCSD module that uses kernel, the integrals arent exactly identical. The non zero terms match but they are shifted in position. So if try to get mp2 density i get two matrices with the same diagonal elements but with different off diagonal parts in the virtual-virtual block due to the difference in the OOVV integral. Both of them might result in same MP2 energy.
I could have used the MP2  module for GHF class but I am actually trying to calculate MP2 density for DHF class which is not available in pyscf for I require the OOVV integral. But due to the difference in OOVV integral i am not getting the exact density matrix which would have matched with the  GHF-MP2 density at the non relativistic limit.

**Comment by sunqm** [comment]
When you call GHF-MP2 or GCCSD with an RHF instance (or DHF, which can be a Kramers-restricted calculation), the orbitals are extended to GHF orbitals with a specific ordering. The orbitals are sorted as spin-up, spin-down, spin-up, spin-down, ...

For closed-shell calculations, the GHF results are very likely equivalent to the RHF results. However, GHF in PySCF does not guarantee a specific ordering of the spin-polarized orbitals, so the orbital order might differ to the order of up, down, up, down...

If you are developing MP2 with DHF orbitals, I'd recommend to overwrite the ao2mo method  https://github.com/pyscf/pyscf/blob/851ef84cf3d6f8b0cfbeacd78018ad816c38cf78/pyscf/mp/gmp2.py#L180 ,  using the `ao2mo.r_outcore` function to generate OOVV integrals.
